jupyterhubutils/authenticator/lsstgithubauth.py

```diff
@@ -0,0 +1,170 @@
+'''LSST-specific Github OAuthenticator class, delegating its LSST-specific
+authentication to its auth_mgr.
+'''
+import json
+import oauthenticator
+from oauthenticator.common import next_page_from_links
+from tornado import gen
+from tornado.httpclient import HTTPRequest, AsyncHTTPClient, HTTPError
+from .. import LSSTMiddleManager
+from ..config import LSSTConfig
+from ..utils import make_logger, sanitize_dict
+
+
+def github_api_headers(access_token):
+    '''Generate API headers for communicating with GitHub.
+    '''
+    return {"Accept": "application/json",
+            "User-Agent": "JupyterHub",
+            "Authorization": "token {}".format(access_token)
+            }
+
+
+class LSSTGitHubOAuthenticator(oauthenticator.GitHubOAuthenticator):
+    enable_auth_state = True
+    login_handler = oauthenticator.GitHubLoginHandler
+    user_groups = []
+
+    def __init__(self, *args, **kwargs):
+        self.log = make_logger()
+        self.log.debug("Creating LSSTGitHubOAuthenticator.")
+        super().__init__(*args, **kwargs)
+        self.log.debug("Superclass authenticator created.")
+        self.lsst_mgr = LSSTMiddleManager(parent=self, config=LSSTConfig())
+
+    @gen.coroutine
+    def authenticate(self, handler, data=None):
+        self.log.info("Authenticating user against GitHub.")
+        userdict = yield super().authenticate(handler, data)
+        try:
+            token = userdict["auth_state"]["access_token"]
+        except (KeyError, TypeError):
+            self.log.warning("Could not extract access token.")
```

Comment:
maybe just return here if you get this error, then you don't have to check if token exists later on